Announce a breaking change of added `dispose()`. (#9397)
flutter/website: src/release/breaking-changes/dispose.md

---
title: Added missed `dispose()` for some disposable objects in Flutter
description: >

>   'dispose()' may fail because of double disposal.

Reviewer: ```suggestion
  'dispose()' might fail because of double disposal.
```

---

## Summary

> Missed 'dispose()' are added for some disposable objects. If some other code
> also invokes 'dispose()' for the objects, and the object is protected from double
> disposal, the second 'dispose()' will fail with error message:

Reviewer 2: **chore (if-minor):** Use [semantic breaks](https://sembr.org/).
```suggestion
Missed 'dispose()' are added for some disposable objects.
If some other code also invokes 'dispose()' for the objects,
and the object is protected from double disposal,
the second 'dispose()' fails with the following error message:
```

Reviewer: Please fix the grammar:
Missed 'dispose()' are added for some disposable objects.
If some other code also invokes 'dispose()' for the objects,

=>
Missing calls to `dispose()` were added for some disposable objects.
If `dispose()` is called elsewhere on the same object

`Once you have called dispose() on a <class name>, it can no longer be used.` 

## Background

> There is a rule that owner of an object should dispose it.

Reviewer: Is this a rule or a convention? If a convention:

```suggestion
The convention is that the owner of an object should dispose of it.
```

> This rule was broken in some places in Flutter Framework: owners were not disposing the disposable objects.
> The issue was fixed by adding `dispose()`. The added `dispose()` can cause failures in debug mode,
> if there is another `dispose()`, and the object is protected from double disposal.

Reviewer 2: **chore (if-minor):** Use [semantic breaks](https://sembr.org/). Condition, then instruction per [GSG](https://developers.google.com/style/sentence-structure). Write in active voice per [GSG](https://developers.google.com/style/voice).
```suggestion
This rule was broken in some places in Flutter Framework:
owners were not disposing the disposable objects.
To fix the issue, `dispose()` was added.
If there is another `dispose()`, and the object is protected from double disposal,
the added `dispose()` can cause failures in debug mode.
```

Reviewer: Please fix the grammar:

=>
This convention was broken in some places:
owners were not disposing the disposable objects.
The issue was fixed by adding a call to `dispose()`.
However, if the object is protected from double disposal,
this can cause failures when running in debug mode
and `dispose()` is called elsewhere on the object.

## Migration guide

> If you got error like `Once you have called dispose() on a <class name>, it can no longer be used.`,
> and the error is raised for your code, 
> update the code to call `dispose()' only in cases when your code created the object.

Reviewer 2: **issue:** Write as shorter sentences per [GSG](https://developers.google.com/style/accessibility#ease-of-reading).
````suggestion
If the following error was raised for your code, 
update the code to call `dispose()' only in cases when your code created the object.

```nocode
Once you have called dispose() on a <class name>, it can no longer be used.
```
````

Reviewer: Please fix the grammar:
If you encounter the following error, then remove
any additional calls in your code that dispose of the object:

```nocode
Once you have called dispose() on a <class name>, it can no longer be used.
```

Author: Shams, in most cases, the code should not be removed, but condition should be added. So, it should be fixed, not removed.

Code before migration:

```dart
x.dispose();
```

Code after migration:

```dart
if (xIsCreatedByMe) {
  x.dispose();
}
```

> If the error is for Flutter code, this means wrong `dispose()` happened first, and the error was raised
> for correct `dispose()` that happened second.
> 
> You can locate the wrong `dispose()` by temporary adding `print(StackTrace.current)`
> to the body of the failed method `dispose`.

Reviewer: It seems like you don't really need this first sentence?
If so, delete. That second sentence is super useful.

```suggestion
If the error occurs in Flutter code, `dispose()` was
called incorrectly the first time.

You can locate an incorrect call to `dispose()` by
temporary calling `print(StackTrace.current)`
in the body of the failed `dispose` method.
```

Author: print(StackTrace.current) makes sense if the error is for Flutter code, otherwise it is not needed.
I updated text to made it clearer.

---
title: Added missing `dispose()` for some disposable objects in Flutter
description: >
  'dispose()' might fail because of double disposal.
---

## Summary

> Missing 'dispose()' are added for some disposable objects.

Reviewer 3: nit: Missing calls to 'dispose()'

> Missing calls to 'dispose()' are added for some disposable objects.

Reviewer 4: For added context it would be nice to list somewhere in this doc examples of where we added previously missing disposes, e.g. "The FooController owned by the BarWidget was missing a dispose" or something like that.
